docs/design-proposals/global-scheduler/GlobalSchedulerArchitectureDesignSpecificationFirstDraft.md

```diff
@@ -0,0 +1,416 @@
+
+In addition to the above requirements, the global scheduling architecture must be able to integrate with any type of clusters on the southbound, such as Kubernetes Clusters, Arktos Clusters, Openstack Clusters, etc.. Priority Scheduling and Fair Scheduling should also be considered in the design.
+
+## 2.2 Key Points of this New Scheduler Architecture Design
+
+Our solution is a new two-level concurrent scheduler architecture built around a globally shared resource pool. This new design aims to achieve both high scalability, low scheduling latency, and high resource utilization while meeting the user-supplied placement constraints through a combination of the following mechanisms:
+
+- shared state lock-free optimistic concurrent scheduling operation
+- multi-dimensional optimization-based scheduling algorithm
+- dynamic geolocation and resource profile-based partition scheme
+- conflict aware VM/Container request distribution and scheduling mechanism
+- smart conflict avoidance approach
+
+## 2.3 High-Level Architecture Design
+
+We divide the scheduling into two levels. The first level Global Scheduler selects the best cluster for the VM/Container request. We then leverage the existing cluster level scheduler to select the best node that will host and run the VM/Container. Our Two-Level Scheduler Architecture is different from what is described in section 1.3.6 since our first level Global Scheduler is not a single monolithic scheduler. It employs a distributed model that consists of multiple Global Schedulers running in a lock-free concurrent mode, thus our new design avoids the issues of scalability bottleneck and low allocation rate. The architecture allows flexible scale-out to the needed global capacity.
+
+In our design, all the Global Schedulers operate in a shared-state manner. That is, they shared one global resource DB which holds the cluster-level resource information of all the clusters. The DB size will be manageable since the DB only holds cluster-level resource information, such as available cluster CPU, available cluster memory, supported VM flavor list, and network topology of clusters, network proximity of clusters, etc.. Unlike the Shared-State Multi-Scheduler Architecture outlined in section 1.3.3, the global resource DB is not replicated by each Global Scheduler. Instead, they share and read/update information from one global resource DB.  In our new design, the Global Schedulers run in an optimistic concurrent way instead of using pessimistic concurrency control, which removes the locking latency.
```

Comment:
Is there a line missing in the high-level architecture shown in the following figure? Is the global scheduler obtaining information from the global resource database? Is this the Dynamic resource info cache?

Reply:
We can discuss this in next meeting. Here is one way to do it---The Dynamic resource info cache is part of the global DB. The difference is that these info does not need persistency so "create" and "update" operation will be much faster. We will leverage the API server to handle the interaction with both DB and cache, similar to what K8S does. That is, the Global Scheduler obtains info from the global DB and cache through API server. API server has cache itself and may return the info from its cache if it is there.  Global Scheduler will have its own cache for the info and may get the info from its own cache.

Reply:
Ok, I think the global scheduler needs to obtain the resource information required for scheduling. Therefore, the global scheduler needs to connect to the database directly or access the database through the API server.

Reply:
The global scheduler will access the DB through the API server, so it will connect directly with the API server, not the DB.